toolchain/lexer/tokenized_buffer_benchmark.cpp

```diff
@@ -0,0 +1,136 @@
+    return std::move(*SourceBuffer::CreateFromFile(fs_, filename_));
+  }
+};
+
+constexpr int NumTokens = 100000;
+
+void BM_ValidKeywords(benchmark::State& state) {
+  absl::BitGen gen;
+  std::string source;
+  llvm::raw_string_ostream os(source);
+  llvm::ListSeparator sep(" ");
+  for (int i = 0; i < NumTokens; ++i) {
+    int token = absl::Uniform<int>(gen, 0, TokenKind::KeywordTokens.size());
+    os << sep << TokenKind::KeywordTokens[token].fixed_spelling();
+  }
+
+  LexerBenchHelper helper(source);
+  for (auto _ : state) {
+    TokenizedBuffer buffer = helper.Lex();
+    benchmark::DoNotOptimize(buffer.has_errors());
+  }
+
+  state.counters["TokenRate"] = benchmark::Counter(
+      NumTokens, benchmark::Counter::kIsIterationInvariantRate);
+}
+BENCHMARK(BM_ValidKeywords);
+
+// clang-format off
+constexpr char IdentifierChars[] = {
+    // Digits:
+    '0', '1', '2', '3', '4', '5', '6', '7', '8', '9',
+    // Upper case:
+    'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O',
+    'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z',
+    // Lower case:
+    'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'M', 'n', 'o',
+    'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z',
+    // Other characters:
+    '_',
+};
+// clang-format on
+
+void BM_ValidIdentifiers(benchmark::State& state) {
+  int min_length = state.range(0);
+  int max_length = state.range(1);
+  absl::BitGen gen;
+  std::string source;
+  llvm::raw_string_ostream os(source);
+  llvm::ListSeparator sep(" ");
+  for (int i = 0; i < NumTokens; ++i) {
+    int length = absl::Uniform<int>(gen, min_length, max_length);
+    os << sep;
+    int id_start = source.size();
+    for (;;) {
+      // Erase any prior attempts to find an identifier.
+      source.resize(id_start);
+      for (int j : llvm::seq<int>(0, length)) {
+        os << IdentifierChars[absl::Uniform<int>(gen, j == 0 ? 10 : 0,
+                                                 sizeof(IdentifierChars))];
+      }
+      // Check if we ended up forming an integer type literal or a keyword, and
+      // try again.
+      auto id = llvm::StringRef(source).substr(id_start);
+      if (llvm::any_of(TokenKind::KeywordTokens, [id](auto token) {
+            return id == token.fixed_spelling();
+          })) {
+        continue;
+      }
+      if (id.consume_front("i") || id.consume_front("u") ||
+          id.consume_front("f")) {
+        if (llvm::all_of(id, [](const char c) { return llvm::isDigit(c); })) {
+          continue;
+        }
+      }
+      break;
+    }
+  }
+
+  LexerBenchHelper helper(source);
+  for (auto _ : state) {
+    TokenizedBuffer buffer = helper.Lex();
+    CARBON_CHECK(!buffer.has_errors()) << helper.DiagnoseErrors();
+  }
+
+  state.counters["TokenRate"] = benchmark::Counter(
+      NumTokens, benchmark::Counter::kIsIterationInvariantRate);
+}
+BENCHMARK(BM_ValidIdentifiers)->Args({3, 5})->Args({3, 16})->Args({12, 64});
```

Comment:
Can you comment with any background on why these ranges? (maybe also noting that these are ranges of identifier sizes)

Reply:
Note, I think even a comment of "// Ranges trying to test a few different potential cases." would be helpful, just so that if someone gets something they want to test more specifically in the future, it's clearer what the standard is to change.